### ISSUE TITLE
Review and maybe optimize performance of `as_next_scheduled_action`

### ISSUE THREAD

**Issue by madeincosmos** [woocommerce/action-scheduler#645]
According to developers of Facebook for WooCommerce, there's a performance problem with the function `as_next_scheduled_action` on large sites. When the scheduled actions table reaches 500,000 records, processing this function takes more than 4s on average. 

In this particular case, Facebook for WooCommerce schedules a feed regeneration job every 15 minutes, so a large number of records is not unusual here. At the same time, `as_next_scheduled_action` is hooked into init, so it adds a significant of delay on every page load:

https://github.com/facebookincubator/facebook-for-woocommerce/issues/1632

As far as I understand, Facebook is going to move away from running this query on init, but it still might be worth checking where this delay could be coming from. I can see that the `store->find_actions()` method called here twice queries the `wp_actionscheduler_actions` table by `hook`, `status`, and `args`, but there's an index created on each of these columns so there's no obvious performance improvement that stands out to me here:

https://github.com/woocommerce/action-scheduler/blob/c899c34a06a987c5143218672c6c9e544c73fed4/classes/data-stores/ActionScheduler_DBStore.php#L214

Scheduled actions are normally cleared after 30 days, so all my test sites have a few thousand records at most. If there's a way to generate a large number of these in bulk, I'll be happy to test it further.

**Comment by barryhughes** [comment]
@madeincosmos I was looking at [PR 688](https://github.com/woocommerce/action-scheduler/pull/688/) and curious if you feel that would effectively resolve this? 

It introduces a new function named `as_has_scheduled_action()` that in many cases—such as testing to see if a recurring action has been setup during `init`—can be used as a faster alternative to `as_next_scheduled_action()`.

Some other high level points (which may also have been surfaced in the linked issues/PRs):

- The consumer can optionally cache the result of a call to `as_next_scheduled_action()` (in a transient, for example). Aside from situations where it is absolutely critical to have a realtime result, that _may_ be preferable to querying on every single request via `init`.
- There are other patterns that can be used to establish recurring scheduled actions, such as setting them up on plugin activation/update or checking to see if they are established/need to be re-established on a separate hourly/daily schedule (which could even run via traditional scehduled events ala _'WP Cron')._

☝🏽 I realize the above points won't be valid in every situation, but wanted to note them even so (also, they are probably just as applicable to the use of a faster alternative like `as_has_scheduled_action()`, if that is introduced via the PR I linked to).

tl;dr — it feels to me like we could close this after merging [PR 688](https://github.com/woocommerce/action-scheduler/pull/688/), but I was interesting in hearing your thoughts.

**Comment by barryhughes** [comment]
Closing per the above.
